Optionally resolve rules by a provided id

benjyw opened the pull request:
This is work towards supporting call-by-name for rules, as spec'd
out in https://github.com/pantsbuild/pants/issues/19730.

This change:

1. Introduces a RuleId struct.
2. Adds an optional RuleId-typed field to DependencyKey.
3. Sets a RuleId on every Task and Intrinsic, of the form
    `path.to.module.function`.
4. Validate that RuleIds are unique.
5. During rule graph construction, when considering candidates
    to satisfy a DependencyKey, if the DependencyKey has a RuleId,
    only consider candidates with that RuleId (there should be only one).
6. A simple test to demonstrate 5.

The idea is that the call-by-name syntax will set the path.to.module.function
as the id on the DependencyKey, which will cause this id to be used to select
the rule and be done with it, instead of the current, complex solving logic.

benjyw commented on src/rust/engine/rule_graph/src/builder.rs:
```diff
@@ -272,6 +275,36 @@ impl<R: Rule> Builder<R> {
+      _ => {
+        invalid_rule_ids.sort();
+        Err(format!(
+          "Multiple rules have these names: {}",
```
The message is a little awkwardly phrased, to keep it simple. But this should never happen, and certainly never happen to end-users.

benjyw commented on src/rust/engine/rule_graph/src/rules.rs:
```diff
@@ -41,6 +66,16 @@ pub struct DependencyKey<T: TypeId> {
 impl<T: TypeId> DependencyKey<T> {
   pub fn new(product: T) -> Self {
     DependencyKey {
+      rule_id: None,
```
Diff algorithm clearly not being helpful, but you get what's going on here.

benjyw commented on src/rust/engine/rule_graph/src/tests.rs:
```diff
@@ -506,27 +539,27 @@ fn reduced_source_roots() {
     Rule::new(
       "Snapshot",
-      "<intrinsic>",
+      "snapshot_from_digest",
```
Had to dedup these now that we validate

benjyw commented on src/rust/engine/src/externs/interface.rs:
```diff
@@ -1014,6 +1014,7 @@ fn session_run_interactive_process(
       &Arc::new(std::sync::atomic::AtomicBool::new(true)),
       core.intrinsics.run(
         &Intrinsic {
+          id: RuleId::new("interactive_process"),
```
This demonstrates that the lookup is now overspecified, and they key could just be the RuleId. But I left that for a followup, to keep this PR focused.

benjyw commented on src/rust/engine/rule_graph/src/tests.rs:
```diff
@@ -661,7 +694,8 @@ fn full_scale_target() {
         DependencyKey::new("InferPythonDependencies"),
         DependencyKey::new("StrippedSourceFiles").provided_params(vec!["SourceFilesRequest"]),
-        DependencyKey::new("PythonModuleOwner").provided_params(vec!["PythonModule"]),
+        DependencyKey::for_known_rule(RuleId::new("map_module_to_address"), "PythonModuleOwner")
```
It made more sense to start incorporating this into the existing tests, rather than duplicate these already elaborate tests.

stuhood replied:
This edit is probably unnecessary given your `by_name_simple` test, but no harm done.

benjyw replied:
I thought it would be good to start integrating this into the more elaborate tests. `by_name_simple` is too simple to be the only test.

benjyw commented:
Looks like we actually do have some rules with non-unique names, due to dynamic rule registration. I will unravel that in a separate PR and then rebase this onto that.

stuhood commented:
Thanks! Will build atop this in the next few days.

huonw commented on src/rust/engine/rule_graph/src/builder.rs:
```diff
@@ -272,6 +275,36 @@ impl<R: Rule> Builder<R> {
+      .rules
+      .values()
+      .flatten()
+      .group_by(|rule| rule.id())
+      .into_iter()
+      .filter_map(|(rule_id, group)| {
+        if group.count() > 1 {
+          Some(rule_id)
+        } else {
+          None
+        }
+      })
```
I note that https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.group_by only groups consecutive equal elements, i.e. `a, b, a` ends up with three groups of size 1.

Is `self.rules.values().flatten()` going to be sorted by `rule.id()`? If not, the docs of https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.duplicates suggest it might handle this well:


```suggestion
      .duplicates()
```

benjyw replied:
Good call! They are not guaranteed to be sorted, and this would explain some weirdness.

stuhood commented on src/rust/engine/rule_graph/src/rules.rs:
```diff
@@ -121,7 +156,13 @@ impl<T: TypeId> DependencyKey<T> {
+    if let Some(rule_id) = &self.rule_id {
+      write!(
+        f,
+        "{}({:?}) -> {}",
```
Maybe include an `...` (ellipses) before or after the provided params, to indicate that we don't know if they are complete?

benjyw replied:
Good idea, done.

stuhood commented on src/rust/engine/src/python.rs:
```diff
@@ -207,6 +207,19 @@ impl Function {
     format!("{module}:{line_no}:{name}")
   }
+
+  /// The function represented as `path.to.module:func_name`. Used to identify a
+  /// rule when called by name.
+  pub fn identifier(&self) -> String {
+    let (module, name) = Python::with_gil(|py| {
+      let val = self.0.to_value();
+      let obj = (*val).as_ref(py);
+      let module: String = externs::getattr(obj, "__module__").unwrap();
+      let name: String = externs::getattr(obj, "__name__").unwrap();
+      (module, name)
+    });
+    format!("{module}:{name}")
+  }
```
Going generic might look like having the "production" `RuleId` implementation be something like:
```rust
enum RuleId {
  Native(&'static str),
  Python(crate::python::Function),
}
```

It _might_ be worthwhile to go generic in this PR if it helps to avoid name collisions? But up to you.

benjyw replied:
The name collisions are all on the Python side, and I've fixed them another way. I don't think this would help with that - the collisions are due to dynamic re-registration of the same function, so the `Function` alone is not enough to disambiguate, we need info from the `@rule` decorator.

benjyw replied:
Not sure we will ever need to go generic here, just a newtyped string seems to be enough.

stuhood reviewed (approved):
Looks great, thanks!

benjyw commented on src/rust/engine/rule_graph/src/builder.rs:
```diff
@@ -362,16 +388,35 @@ impl<R: Rule> Builder<R> {
-            candidates.push(Node::Param(dependency_key.product()));
-          }
+          if let Some(rule_id) = &dependency_key.rule_id {
+            if let Some(rules) = self.rules.get(&dependency_key.product()) {
+              candidates.extend(rules.iter().filter_map(|r| {
+                if r.id() == rule_id {
+                  Some(Node::Rule(r.clone()))
+                } else {
+                  None
+                }
+              }));
+            };
```
Yeah, good idea. Mostly for obviousness, since we're only iterating over the rules with a matching dependency key, which should be small. But, yeah.

benjyw commented:
OK, two new commits to review:

1. Changes to how the RuleId is generated for Python @rules: Turns out we already had a `canonical_name` property that we set but didn't really use, except for display. We now use this as the RuleId, and we ensure it is unique even when rules are dynamically registered multiple times, via a suffix mechanism.
2. Addresses the code review comments.

PTAL, thanks!

benjyw commented:
@stuhood please PTAL? There are some substantial changes.

stuhood commented on src/python/pants/engine/rules.py:
```diff
@@ -244,6 +253,8 @@ def rule_decorator(func, **kwargs) -> Callable:
         pants_rules[effective_name] = func
     else:
         prev_func = pants_rules[effective_name]
+        # TODO: Remove this check? We also check for uniqueness of effective_name in Rust code,
+        #  and that check is more comprehensive, as it also covers intrinsics.
```
This catches the case where two rules with the same name are declared in a module (which Python would otherwise allow). In that case, only one would actually end up accessible in the module, and the Rust-side validation wouldn't catch it.

benjyw replied:
Right, I have removed this TODO.

stuhood commented on src/python/pants/engine/rules.py:
```diff
@@ -217,9 +217,18 @@ def rule_decorator(func, **kwargs) -> Callable:
     # Set a default canonical name if one is not explicitly provided to the module and name of the
-    # function that implements it. This is used as the workunit name.
+    # function that implements it, plus an optional suffix. This is used as the workunit name.
+    # The suffix is a convenient way to disambiguate multiple rules registered dynamically from the
+    # same static code (by overriding the inferred param types in the @rule decorator).
```
I still think that making the rule id generic would be cleaner, but not blocking.

benjyw replied:
Should be easy to do later if we need to. But it really is just a string.
